python/lsst/pipe/base/quantum_provenance_graph.py

```diff
@@ -0,0 +1,961 @@
+
+
+class QuantumInfo(TypedDict):
+    """Information about a quantum across all run collections.
+
+    Used to annotate the networkx node dictionary.
+    """
+
+    data_id: DataCoordinate
+    """The data_id of the quantum.
+    """
+
+    runs: dict[str, QuantumRun]
+    """All run collections associated with the quantum.
+    """
+
+    status: QuantumInfoStatus
+    """The overall status of the quantum. Note that it is impossible to exit a
+    wonky state.
+    """
+
+    recovered: bool
+    """The quantum was originally not successful but was ultimately successful.
+    """
+
+    messages: list[str]
+    """Diagnostic messages to help disambiguate wonky states.
+    """
+
+    log: DatasetKey
+    """The `DatasetKey` which can be used to access the log associated with the
+    quantum.
+    """
+
+    metadata: DatasetKey
+    """The `DatasetKey` which can be used to access the metadata for the
+    quantum.
+    """
+
+
+class DatasetRun(pydantic.BaseModel):
+    """Information about a dataset in a given run collection."""
+
+    id: uuid.UUID
+    """The dataset ID associated with the dataset in a specific run.
+    """
+
+    produced: bool = False
+    """Whether the specific run produced the dataset.
+    """
+
+    published: bool = False
+    """Whether this dataset was published in the final output collection.
+    """
+
+    @pydantic.model_validator(mode="after")
+    def _validate(self) -> DatasetRun:
+        """Validate the model for `DatasetRun` by asserting that no published
+        `DatasetRun` is also not produced (this should be impossible).
+        """
+        assert not (self.published and not self.produced)
+        return self
+
+
+DatasetInfoStatus: TypeAlias = Literal["published", "unpublished", "predicted_only", "unsuccessful", "cursed"]
+
+
+class DatasetInfo(TypedDict):
+    """Information about a given dataset across all runs.
+
+    Used to annotate the networkx node dictionary.
+    """
+
+    data_id: DataCoordinate
+    """The data_id of the quantum.
+    """
+
+    runs: dict[str, DatasetRun]
+    """All runs associated with the dataset.
+    """
+
+    status: DatasetInfoStatus
+    """Overall status of the dataset.
+    """
+
+    messages: list[str]
+    """Diagnostic messages to help disambiguate cursed states.
+    """
+
+
+class UnsuccessfulQuantumSummary(pydantic.BaseModel):
+    """A summary of all relevant information on an unsuccessful quantum."""
+
+    data_id: dict[str, DataIdValue]
+    """The data_id of the unsuccessful quantum.
+    """
+    runs: dict[str, QuantumRunStatus]
+    """A dictionary including the `QuantumRunStatus` of each run associated
+    with an attempt to process the unsuccessful quantum.
+    """
+    messages: list[str]
+    """Any messages associated with the unsuccessful quantum (any clues as to
+    why the quantum may be in a `failed` or `wonky` state).
+    """
```

Comment:
I did think about this-- it might end up on a later ticket. These will either be error messages in the case of failed quanta or reasons why they're wonky-- a short list to be sure.